firecracker-control/local.go

```diff
@@ -311,3 +343,93 @@ func (s *local) getVM(id string) (*instance, error) {
 
 	return instance, nil
 }
+
+// PatchStubDrive will replace a stub drive in the queue with a new path.
+func (s *local) PatchStubDrive(ctx context.Context, req *proto.PatchStubDriveRequest) (*proto.PatchStubDriveResponse, error) {
+	if s.stubDriveIndex >= len(s.drives) {
+		return nil, ErrStubDrivesExhausted
+	}
+
+	d := s.drives[s.stubDriveIndex]
+	d.PathOnHost = &req.PathOnHost
+
+	if d.DriveID == nil {
+		// this should never happen, but we want to ensure that we never nil
+		// dereference
+		return nil, ErrDriveIDNil
+	}
+
+	s.drives[s.stubDriveIndex] = d
+
+	instance, err := s.getVM(req.GetVMID())
+	if err != nil {
+		return nil, err
+	}
+
+	err = instance.machine.UpdateGuestDrive(ctx, firecracker.StringValue(d.DriveID), req.PathOnHost)
+	if err != nil {
+		return nil, err
+	}
+
+	s.stubDriveIndex++
+	return &proto.PatchStubDriveResponse{
+		DriveID: *d.DriveID,
+	}, nil
+}
+
+// setupStubDrive will create a flat file with the drive ID as its contents.
+// The file is padded with null bytes to a minimum size of minFCDriveImageSize.
+// This is due to Firecracker minimum drive image size limits.
+func setupStubDrive(builder firecracker.DrivesBuilder, dir, driveID string) (firecracker.DrivesBuilder, error) {
+	path := filepath.Join(dir, driveID)
+
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0600)
+	if err != nil {
+		return builder, err
+	}
+	defer f.Close()
+
+	if _, err = f.WriteString(driveID); err != nil {
+		return builder, err
+	}
+
+	// Firecracker will not show any drives smaller than 512 bytes. If the drive
+	// size is not a multiple of 512, then those bytes will not be visible to
+	// firecracker.
+	if err := os.Truncate(path, minFCDriveImageBytes); err != nil {
+		return builder, err
+	}
+
+	builder = builder.AddDrive(path, false, func(drive *models.Drive) {
+		drive.DriveID = firecracker.String(driveID)
+	})
+
+	return builder, nil
+}
+
+// isStubDrive will iterate through a list of drives to check if the path on
+// the current drive exists in the cached drives we have. Since stub drives
+// will be replaced with actual drives at some point, drives that are not in
+// the list means that a stub drive is being replaced.
+//
+// TODO: Optimize? I imagine n is small, currently max is 15, so maybe not.
+func (s *local) isStubDrive(drive *proto.FirecrackerDrive) bool {
```

Comment:
Is this `isStubDrive` method on the local struct used anywhere? I can't find it (but that may just be me)